genesis/engine/entities/particle_entity.py

```diff
@@ -301,22 +301,32 @@ def reset_grad(self):
 
     def set_velocity(self, vel):
         """
-        Accepted tensor shape: (3,) or (self._n_particles, 3).
+        Accepted tensor shape: (3,) or (self._n_particles, 3) or (self._sim._B, self._n_particles, 3).
         """
+
         self._assert_active()
         if self.sim.requires_grad:
             gs.logger.warning(
-                "Manually setting particle velocities. This is not recommended and could break gradient flow."
+                "Manally setting element velocities. This is not recommended and could break gradient flow."
             )
 
         vel = to_gs_tensor(vel)
 
         if len(vel.shape) == 1:
             assert vel.shape == (3,)
-            self._tgt["vel"] = torch.tile(vel, [self._n_particles, 1])
+            self._tgt["vel"] = torch.tile(vel, [self._sim._B, self._n_particles, 1])
 
         elif len(vel.shape) == 2:
-            assert vel.shape == (self._n_particles, 3)
+            assert self.n_particles != n_groups
+            if vel.shape == (self._n_particles, 3):
+                self._tgt["vel"] = torch.tile(vel[None], [self._sim._B, 1, 1])
```

Comment:
Prefer `unsqueeze` for torch tensors or `np.newaxis` for numpy arrays over `None` if possible.